hyperon_das/traverse_engines.py

```diff
@@ -0,0 +1,197 @@
+        return self._cursor
+
+    def get_links(self, **kwargs) -> ListIterator:
+        incoming_links = self._get_incoming_links()
+
+        if not incoming_links:
+            return ListIterator([])
+
+        filtered_links = self._filter_links(
+            links=incoming_links,
+            link_type=kwargs.get('link_type'),
+            cursor_position=kwargs.get('cursor_position'),
+            target_type=kwargs.get('target_type'),
+            handles_only=True,
+        )
+
+        return ListIterator(filtered_links)
+
+    def get_neighbors(self, **kwargs) -> List[str]:
+        filtered_links_iterator = self.get_links(
+            link_type=kwargs.get('link_type'),
+            target_type=kwargs.get('target_type')
+        )
+
+        targets_set = {target for link in filtered_links_iterator for target in link['targets']}
+
+        if not targets_set:
+            return []
+
+        targets_set.discard(self._cursor)
+
+        return list(targets_set)
+
+    def follow_link(self, **kwargs) -> None:
+        filtered_links_iterator = self.get_links(
+            link_type=kwargs.get('link_type'),
+            target_type=kwargs.get('target_type')
+        )
+
+        filtered_links = [link for link in filtered_links_iterator]
+
+        if not filtered_links:
+            return
+
+        unique_path = kwargs.get('unique_path', False)
+
+        if unique_path and len(filtered_links) > 1:
+            raise MultiplePathsError(
+                message='Unable to follow the link. More than one path found',
+                details=f'{len(filtered_links)} paths',
+            )
+
+        link = choice(filtered_links)
+
+        if self._cursor in link['targets']:
+            link['targets'].remove(self._cursor)
+
+        handle = choice(link['targets'])
+
+        self.goto(handle)
+
+
+class DocumentTraverseEngine(TraverseEngine):
+    def get(self) -> Dict[str, Any]:
+        return self.das.get_atom(self._cursor)
+
+    def get_links(self, **kwargs) -> ListIterator:
+        incoming_links = self._get_incoming_links()
+
+        if not incoming_links:
+            return ListIterator([])
+
+        filtered_links = self._filter_links(
+            links=incoming_links,
+            link_type=kwargs.get('link_type'),
+            cursor_position=kwargs.get('cursor_position'),
+            target_type=kwargs.get('target_type'),
+            custom_filter=kwargs.get('filter'),
+        )
+
+        return ListIterator(filtered_links)
+
+    def get_neighbors(self, **kwargs) -> List[Dict[str, Any]]:
+        filtered_links_iterator = self.get_links(
+            link_type=kwargs.get('link_type'),
+            target_type=kwargs.get('target_type'),
+            filter=kwargs.get('filter'),
+        )
+
+        targets_document = []
+        for link in filtered_links_iterator:
+            for document in link['targets_document']:
+                if self._cursor == document['handle'] or document in targets_document:
```

Comment:
The check for document is target_document is insanely expensive. Please redesign.